engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java

```diff
@@ -164,28 +171,41 @@ void replayNextEvent() {
       }
 
     } catch (final RuntimeException e) {
-      final var replayException =
-          new ProcessingException("Unable to replay record", currentEvent, metadata, e);
-      recoveryFuture.completeExceptionally(replayException);
+      recoveryFuture.completeExceptionally(new RuntimeException("Failed to replay records", e));
     }
   }
 
-  private void replayEvent(final LoggedEvent currentEvent) {
-    if (!eventFilter.applies(currentEvent)) {
-      onRecordReplayed(currentEvent);
-      return;
+  private boolean tryToReplayBatch(final Batch batch) throws Exception {
+    final boolean onRetry = zeebeDbTransaction != null;
+    if (onRetry) {
+      zeebeDbTransaction.rollback();
+      // reading the whole batch from the beginning again
+      batch.head();
     }
 
-    readMetadata(currentEvent);
-    final var currentTypedEvent = readRecordValue(currentEvent);
+    zeebeDbTransaction = transactionContext.getCurrentTransaction();
+    zeebeDbTransaction.run(
+        () -> {
+          batch.forEachRemaining(this::replayEvent);
+
+          lastProcessedPositionState.markAsProcessed(lastSourceEventPosition);
```

Comment:
:thought_balloon: This atomically updates the lastProcessedPosition in the state. But for snapshotting, we use the inmemory `lastSourceEventPosition` (`StreamProcessor:getLastProcessedPositionAsync`) which is not updated atomically after a batch is processed. Instead it is updated after each event is replayed. Would this be a problem? It is used to decide which position can be compacted.

Reply:
Good point. It's better to set the position after the transaction is committed :+1: